// ng-textflow/esm5/ng-textflow/ng-textflow.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, ViewChild, ViewContainerRef, Input } from '@angular/core';
import { CreateFlowService } from './create-flow.service';
var NgTextflowComponent = /** @class */ (function () {
    function NgTextflowComponent(nodeService) {
        this.nodeService = nodeService;
        this.nodeStyles = {};
        this.firstOnTop = true;
        nodeService.firstOnTop = this.firstOnTop;
    }
    Object.defineProperty(NgTextflowComponent.prototype, "content", {
        get: /**
         * @return {?}
         */
        function () { return this._content; },
        set: /**
         * @param {?} text
         * @return {?}
         */
        function (text) {
            this._content = text;
            // Clear the nodes out of the NodeComponent[] array.
            this.nodeService.clearNodes();
            this.viewContainer.clear();
            this.nodeService.createFlow(this.content, this.viewContainer, this.nodeStyles, this.showPageNumbers, this.isOverlaid);
        },
        enumerable: true,
        configurable: true
    });
    NgTextflowComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ng-textflow',
                    template: " \n    <ng-container #viewContainer></ng-container>\n  "
                }] }
    ];
    /** @nocollapse */
    NgTextflowComponent.ctorParameters = function () { return [
        { type: CreateFlowService }
    ]; };
    NgTextflowComponent.propDecorators = {
        viewContainer: [{ type: ViewChild, args: ['viewContainer', { read: ViewContainerRef },] }],
        nodeStyles: [{ type: Input }],
        firstOnTop: [{ type: Input }],
        showPageNumbers: [{ type: Input }],
        isOverlaid: [{ type: Input }],
        content: [{ type: Input }]
    };
    return NgTextflowComponent;
}());
export { NgTextflowComponent };
if (false) {
    /** @type {?} */
    NgTextflowComponent.prototype.viewContainer;
    /** @type {?} */
    NgTextflowComponent.prototype.nodeStyles;
    /** @type {?} */
    NgTextflowComponent.prototype.firstOnTop;
    /** @type {?} */
    NgTextflowComponent.prototype.showPageNumbers;
    /** @type {?} */
    NgTextflowComponent.prototype.isOverlaid;
    /** @type {?} */
    NgTextflowComponent.prototype._content;
    /** @type {?} */
    NgTextflowComponent.prototype.nodeService;
}
