// ng-textflow/esm5/ng-textflow/ng-textflow.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgTextflowComponent } from './ng-textflow.component';
import { NodeComponent } from './node.component';
import { CreateFlowService } from './create-flow.service';
var NgTextflowModule = /** @class */ (function () {
    function NgTextflowModule() {
    }
    /**
     * @return {?}
     */
    NgTextflowModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: NgTextflowModule,
            providers: [CreateFlowService]
        };
    };
    NgTextflowModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule],
                    declarations: [NgTextflowComponent, NodeComponent],
                    entryComponents: [NodeComponent],
                    exports: [NgTextflowComponent]
                },] }
    ];
    return NgTextflowModule;
}());
export { NgTextflowModule };
