// ng-textflow/fesm2015/ng-textflow.js
import { Component, ElementRef, ViewChild, Renderer2, HostBinding, Injectable, Injector, ComponentFactoryResolver, ViewContainerRef, Input, NgModule } from '@angular/core';
import { Subject, interval, BehaviorSubject } from 'rxjs';
import { debounce } from 'rxjs/operators';
import { CommonModule } from '@angular/common';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class NodeComponent {
    /**
     * @param {?} renderer
     * @param {?} elementRef
     */
    constructor(renderer, elementRef) {
        this.renderer = renderer;
        this.elementRef = elementRef;
        this._heading = '';
        this._isVisible = false;
        this._pageNumberAtTop = false;
        this.showHeading = true;
        // Get host properties to dynamically change.
        this.hostPosition = '';
        this.hostVisibility = 'hidden';
        // Dynamic styles.
        this.numberStyle = {};
        this.headingStyle = {};
        this.nodeStyle = {};
        this.hostStyle = {};
        this.defaultStyles = {
            hostStyle: {},
            contentStyle: {
                'height': '9em',
                'width': '200px',
                'fontSize': '11px',
                'wordSpacing': '2px',
                'textAlign': 'justify',
                'lineHeight': '1.1em',
                'leftAlignLast': 'justified'
            },
            headingStyle: {
                'font-size': '14px',
                'height': '25px'
            },
            numberStyle: {
                'font-size': '10px',
            }
        };
        this.overflowSubject = new Subject();
        this.overflow = /** @type {?} */ (this.overflowSubject);
        this.adding = false;
        this.finishing = false;
        this.trimming = false;
        this.count = 100;
    }
    ;
    /**
     * @return {?}
     */
    ngOnInit() {
        // Some initialisation.
        this.dummyText = this.text;
        this.trimming = true;
        // We use the mutation observer to see when our dummy text change has happened.
        this.changes = new MutationObserver((mutations) => this.manageState());
        this.changes.observe(this.dummyDivRef.nativeElement, { attributes: true, childList: true, characterData: true });
        // Kick off the state machine here, but with a delay for the first node,
        // since the initial loading of the component causes some delay in other initialisation. (I think).
        setTimeout(() => {
            // Wait till index is resolved.
            // We need to pause for the first node since the loading is lazy?
            if (this.index >= 99) {
                setTimeout(() => this.manageState(), 200); // TODO: find a event driven method here.
            }
            else {
                // The following nodes do not need this delay.
                this.manageState();
            }
        });
    }
    /**
     * @return {?}
     */
    get dummyText() { return this.dummyDivRef.nativeElement.textContent; }
    /**
     * @param {?} content
     * @return {?}
     */
    set dummyText(content) { this.dummyDivRef.nativeElement.textContent = content; }
    /**
     * @return {?}
     */
    get isVisible() { return this._isVisible; }
    /**
     * @param {?} value
     * @return {?}
     */
    set isVisible(value) {
        this._isVisible = value;
        this.hostVisibility = 'visible';
    }
    /**
     * @return {?}
     */
    get index() { return this._index; }
    /**
     * @param {?} index
     * @return {?}
     */
    set index(index) {
        this.pageNumber = 100 - index;
        setTimeout(() => {
            this._index = Math.abs(index);
            this.renderer.setStyle(this.elementRef.nativeElement, 'z-index', index);
        });
    }
    /**
     * @return {?}
     */
    get width() { return parseInt(this.mask.nativeElement.style.width.replace(/\D/g, '')); }
    /**
     * @param {?} width
     * @return {?}
     */
    set width(width) { this.renderer.setStyle(this.mask.nativeElement, 'width', `${width}px`); }
    /**
     * @return {?}
     */
    get heading() { return this._heading; }
    /**
     * @param {?} heading
     * @return {?}
     */
    set heading(heading) {
        this._heading = heading === '' ? this._heading = ' ' : this._heading = heading;
        NodeComponent.headingChangedSubject.next();
    }
    /**
     * @return {?}
     */
    get pageNumberAtTop() { return this._pageNumberAtTop; }
    /**
     * @param {?} value
     * @return {?}
     */
    set pageNumberAtTop(value) {
        this._pageNumberAtTop = value;
        this.numberStyle['padding-bottom'] = this.pageNumberAtTop ? '0px' : `${this.padding}px`;
        this.numberStyle['padding-top'] = this.pageNumberAtTop ? `${this.padding}px` : '0px';
    }
    /**
     * @param {?} style
     * @return {?}
     */
    applyStyle(style) {
        this.nodeStyle = style.contentStyle ? this.convertStyleToPx(style.contentStyle) : this.convertStyleToPx(this.defaultStyles.contentStyle);
        this.numberStyle = style.numberStyle ? style.numberStyle : this.defaultStyles.numberStyle;
        this.hostStyle = style.hostStyle ? style.hostStyle : this.defaultStyles.hostStyle;
        // Let's manipulate the padding style that was added to manage top or bottom placed numbering.
        if (this.numberStyle.padding > '') {
            this.padding = parseInt(this.numberStyle.padding.replace(/\D/g, ''));
        }
        else {
            this.padding = 16;
        }
        this.numberStyle['padding-left'] = `${parseInt(this.nodeStyle.width.replace(/\D/g, '')) / 2}px`;
        this.headingStyle = style.headingStyle ? style.headingStyle : this.defaultStyles.headingStyle;
        this.headingStyle.width = this.nodeStyle.width;
        this.renderer.setStyle(this.elementRef.nativeElement, 'width', style.width);
        this.nodeStyle.overflow = 'hidden';
        this.width = parseInt(this.nodeStyle.width.replace(/\D/g, ''));
        if (this.isOverlaid) {
            this.hostPosition = 'absolute';
        }
    }
    /**
     * @param {?} style
     * @return {?}
     */
    convertStyleToPx(style) {
        /** @type {?} */
        let tempStyle = style;
        /** @type {?} */
        let fontSizeSuffix = tempStyle['fontSize'].replace(/[\d\.]/g, '');
        /** @type {?} */
        let fontSize = parseFloat(tempStyle['fontSize'].replace(/[A-Za-z]/g, ''));
        for (let key in tempStyle) {
            // Not we need to check each value only, for em, pt, px etc..
            if ((/** @type {?} */ (tempStyle[key])).indexOf('em') > 0) {
                /** @type {?} */
                let suffix = tempStyle[key].replace(/[\d\.]/g, '');
                switch (suffix) {
                    case 'em': {
                        /** @type {?} */
                        let value = parseFloat(tempStyle[key].replace(/[A-Za-z]/g, ''));
                        tempStyle[key] = `${Math.round(value * fontSize)}px`;
                    }
                }
            }
        }
        return tempStyle;
    }
    /**
     * @return {?}
     */
    parseLineHeight() {
        /** @type {?} */
        let lineHeight = this.textDivRef.nativeElement.style.lineHeight;
        return parseInt(lineHeight.replace(/\D/g, ''));
    }
    /**
     * @return {?}
     */
    compareDivs() {
        /** @type {?} */
        let lineHeight = this.parseLineHeight();
        /** @type {?} */
        let dummyLines = Math.floor(this.dummyDivRef.nativeElement.clientHeight / lineHeight);
        /** @type {?} */
        let contentLines = Math.floor(this.textDivRef.nativeElement.clientHeight / lineHeight);
        /** @type {?} */
        let dummyTextWordCount = this.dummyText.split(' ').length;
        return {
            'lines': dummyLines,
            'diffLines': dummyLines - contentLines,
            'words': dummyTextWordCount
        };
    }
    /**
     * @return {?}
     */
    manageState() {
        // Protect against infinite loop.
        if (this.count-- <= 0) {
            this.trimming = false;
            this.adding = false;
            this.finishing = false;
            console.log('ERROR - loop count > 100');
        }
        if (this.trimming) {
            this.trim();
        }
        else if (this.adding) {
            this.addword();
        }
        else if (this.finishing) {
            this.finish();
        }
    }
    /**
     * @return {?}
     */
    trim() {
        /** @type {?} */
        let comps = this.compareDivs();
        /** @type {?} */
        let wordsToRemove = Math.floor(comps.words / comps.lines) * (comps.diffLines + 1);
        if (wordsToRemove > 0) {
            this.dummyText = this.dummyText.split(' ').slice(0, comps.words - wordsToRemove).join(' ');
        }
        else {
            this.trimming = false;
            this.adding = true;
            this.addword();
        }
    }
    /**
     * @return {?}
     */
    addword() {
        /** @type {?} */
        let comps = this.compareDivs();
        if (comps.diffLines <= -1) {
            if (this.dummyText.length < this.text.length) {
                this.dummyText = this.dummyText + ' ' + this.text.split(' ')[comps.words];
            }
            else {
                // This is the last node.
                this.renderer.setStyle(this.textDivRef.nativeElement, 'textAlignLast', 'left');
                NodeComponent.finishedSubject.next();
            }
        }
        else {
            // Take off the last word again.
            this.dummyText = this.dummyText.split(' ').slice(0, comps.words - 1).join(' ');
            this.adding = false;
            this.finishing = true;
        }
    }
    /**
     * @return {?}
     */
    finish() {
        /** @type {?} */
        let comps = this.compareDivs();
        /** @type {?} */
        let lastword = this.dummyText.split(' ');
        /** @type {?} */
        let overflowArray = this.text.split(' ');
        /** @type {?} */
        let overflow = overflowArray.slice(this.dummyText.split(' ').length).join(' ');
        this.text = this.dummyText;
        this.finishing = false;
        this.overflowSubject.next(overflow);
    }
}
NodeComponent.finishedSubject = new Subject();
NodeComponent.finished = /** @type {?} */ (NodeComponent.finishedSubject);
NodeComponent.headingChangedSubject = new Subject();
NodeComponent.headingChanged = NodeComponent.headingChangedSubject.pipe(debounce(() => interval(10)));
NodeComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-node',
                template: `
    <div class="mask" #maskTag [ngStyle]="hostStyle">
      <div *ngIf="showPageNumber && pageNumberAtTop" [ngStyle]="numberStyle">
        {{pageNumber}}
      </div>  
      <div *ngIf="showHeading" [ngStyle]="headingStyle">
        {{heading}}
      </div>
      <div #textref [ngStyle]="nodeStyle" class="content">
        {{text}}
        <div #dummy class="dummy" >{{ dummyText }}</div>
      </div>
      <div *ngIf="showPageNumber && !pageNumberAtTop" [ngStyle]="numberStyle">
        {{pageNumber}}
      </div>
    </div>
`,
                styles: [`
    .content {
      position: relative;
      text-align-last: justify;
      cursor: default;
    }
    .mask {
      display block;
      overflow: hidden;
    }
    :host {
      display: block;
      overflow: hidden;
      user-select: none;
    }
    .dummy {
      position: absolute;
      background-color: rgba(155, 255, 155, 0.4);
      left: 0px;
      top: 0px;
      visibility: hidden;
    }
  `]
            }] }
];
/** @nocollapse */
NodeComponent.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef }
];
NodeComponent.propDecorators = {
    textDivRef: [{ type: ViewChild, args: ['textref',] }],
    dummyDivRef: [{ type: ViewChild, args: ['dummy',] }],
    mask: [{ type: ViewChild, args: ['maskTag',] }],
    hostPosition: [{ type: HostBinding, args: ['style.position',] }],
    hostVisibility: [{ type: HostBinding, args: ['style.visibility',] }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class CreateFlowService {
    /**
     * @param {?} injector
     * @param {?} resolver
     */
    constructor(injector, resolver) {
        this.injector = injector;
        this.resolver = resolver;
        this.nodeList = new Array();
        this.firstOnTop = true;
        this.overFlowSubscriptions = [];
        this.nodesSubject = new BehaviorSubject(null);
        this.nodes = /** @type {?} */ (this.nodesSubject);
        this.nodeFactory = this.resolver.resolveComponentFactory(NodeComponent);
        // When all nodes are finished generating, do some housework.
        this.finishedSubscription = NodeComponent.finished.subscribe(() => {
            this.nodeList.forEach(node => node.isVisible = true); // Make all visible once complete.
            // If headings are modified externally, then update sibling nodes to maintain consistent formating.
            this.headingChangedSubscription = NodeComponent.headingChanged.subscribe(() => {
                /** @type {?} */
                let count = 0;
                this.nodeList.forEach(node => {
                    count = node.heading <= ' ' ? count + 1 : count - 1;
                });
                // Loop again to update the visibility of the headings.
                this.nodeList.forEach(node => {
                    node.showHeading = count === this.nodeList.length ? false : true;
                });
            });
            // Make list of nodes available once all are complete.
            this.nodesSubject.next(this.nodeList);
        });
    }
    /**
     * @return {?}
     */
    clearNodes() {
        this.nodeList = [];
    }
    /**
     * @param {?} textPassage
     * @param {?} location
     * @param {?} style
     * @param {?} showPageNumbers
     * @param {?} isOverlaid
     * @return {?}
     */
    createFlow(textPassage, location, style, showPageNumbers, isOverlaid) {
        /** @type {?} */
        let node = this.createNode(textPassage, location, style, showPageNumbers, isOverlaid);
        // Create a new node if there is any overflow text from last node generated.
        // Last node will know when it is the last one, and will
        this.overFlowSubscriptions.push(node.overflow.subscribe(overflowText => {
            if (overflowText) {
                this.createFlow(overflowText, location, style, showPageNumbers, isOverlaid);
            }
        }));
    }
    /**
     * @param {?} content
     * @param {?=} location
     * @param {?=} style
     * @param {?=} showPageNumbers
     * @param {?=} isOverlaid
     * @return {?}
     */
    createNode(content, location, style, showPageNumbers, isOverlaid) {
        /** @type {?} */
        let nodeRef;
        if (location || location instanceof ViewContainerRef) {
            nodeRef = location.createComponent(this.nodeFactory, undefined, this.injector);
        }
        else {
            nodeRef = this.nodeFactory.create(this.injector);
        }
        // Update the node
        nodeRef.instance.isOverlaid = isOverlaid;
        nodeRef.instance.showPageNumber = showPageNumbers;
        nodeRef.instance.text = content;
        if (style) {
            nodeRef.instance.applyStyle(style);
        }
        // Add node to list of nodes and update its index value.
        this.nodeList.push(nodeRef.instance);
        nodeRef.instance.index = this.firstOnTop ? 100 - this.nodeList.length : this.nodeList.length;
        return nodeRef.instance;
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.finishedSubscription.unsubscribe();
        this.headingChangedSubscription.unsubscribe();
        this.overFlowSubscriptions.forEach(subscription => {
            subscription.unsubscribe();
        });
    }
}
CreateFlowService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
CreateFlowService.ctorParameters = () => [
    { type: Injector },
    { type: ComponentFactoryResolver }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
// A basic component that uses the CreateFlowService to generate and display the nodes
// based on the content provided in these inputs.
class NgTextflowComponent {
    /**
     * @param {?} nodeService
     */
    constructor(nodeService) {
        this.nodeService = nodeService;
        this.nodeStyles = {};
        this.firstOnTop = true;
        nodeService.firstOnTop = this.firstOnTop;
    }
    /**
     * @return {?}
     */
    get content() { return this._content; }
    /**
     * @param {?} text
     * @return {?}
     */
    set content(text) {
        this._content = text;
        // Clear the nodes out of the NodeComponent[] array.
        this.nodeService.clearNodes();
        this.viewContainer.clear();
        this.nodeService.createFlow(this.content, this.viewContainer, this.nodeStyles, this.showPageNumbers, this.isOverlaid);
    }
}
NgTextflowComponent.decorators = [
    { type: Component, args: [{
                selector: 'ng-textflow',
                template: ` 
    <ng-container #viewContainer></ng-container>
  `
            }] }
];
/** @nocollapse */
NgTextflowComponent.ctorParameters = () => [
    { type: CreateFlowService }
];
NgTextflowComponent.propDecorators = {
    viewContainer: [{ type: ViewChild, args: ['viewContainer', { read: ViewContainerRef },] }],
    nodeStyles: [{ type: Input }],
    firstOnTop: [{ type: Input }],
    showPageNumbers: [{ type: Input }],
    isOverlaid: [{ type: Input }],
    content: [{ type: Input }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class NgTextflowModule {
    /**
     * @return {?}
     */
    static forRoot() {
        return {
            ngModule: NgTextflowModule,
            providers: [CreateFlowService]
        };
    }
}
NgTextflowModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule],
                declarations: [NgTextflowComponent, NodeComponent],
                entryComponents: [NodeComponent],
                exports: [NgTextflowComponent]
            },] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

export { NgTextflowModule, NodeComponent, CreateFlowService, NgTextflowComponent as ɵa };
