// ng-textflow/esm2015/ng-textflow/ng-textflow.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, ViewChild, ViewContainerRef, Input } from '@angular/core';
import { CreateFlowService } from './create-flow.service';
// A basic component that uses the CreateFlowService to generate and display the nodes
// based on the content provided in these inputs.
export class NgTextflowComponent {
    /**
     * @param {?} nodeService
     */
    constructor(nodeService) {
        this.nodeService = nodeService;
        this.nodeStyles = {};
        this.firstOnTop = true;
        nodeService.firstOnTop = this.firstOnTop;
    }
    /**
     * @return {?}
     */
    get content() { return this._content; }
    /**
     * @param {?} text
     * @return {?}
     */
    set content(text) {
        this._content = text;
        // Clear the nodes out of the NodeComponent[] array.
        this.nodeService.clearNodes();
        this.viewContainer.clear();
        this.nodeService.createFlow(this.content, this.viewContainer, this.nodeStyles, this.showPageNumbers, this.isOverlaid);
    }
}
NgTextflowComponent.decorators = [
    { type: Component, args: [{
                selector: 'ng-textflow',
                template: ` 
    <ng-container #viewContainer></ng-container>
  `
            }] }
];
/** @nocollapse */
NgTextflowComponent.ctorParameters = () => [
    { type: CreateFlowService }
];
NgTextflowComponent.propDecorators = {
    viewContainer: [{ type: ViewChild, args: ['viewContainer', { read: ViewContainerRef },] }],
    nodeStyles: [{ type: Input }],
    firstOnTop: [{ type: Input }],
    showPageNumbers: [{ type: Input }],
    isOverlaid: [{ type: Input }],
    content: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    NgTextflowComponent.prototype.viewContainer;
    /** @type {?} */
    NgTextflowComponent.prototype.nodeStyles;
    /** @type {?} */
    NgTextflowComponent.prototype.firstOnTop;
    /** @type {?} */
    NgTextflowComponent.prototype.showPageNumbers;
    /** @type {?} */
    NgTextflowComponent.prototype.isOverlaid;
    /** @type {?} */
    NgTextflowComponent.prototype._content;
    /** @type {?} */
    NgTextflowComponent.prototype.nodeService;
}
